controlplane/kubeadm/api/v1alpha4/kubeadmcontrolplane_validation.go

```diff
@@ -0,0 +1,211 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha4
+
+import (
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"sigs.k8s.io/cluster-api/util/container"
+	"sigs.k8s.io/cluster-api/util/version"
+)
+
+func (s *KubeadmControlPlaneSpec) validate(namespace string) (allErrs field.ErrorList) {
```

Comment:
One more suggestion about the method,  since we are refactoring it,  let's. break this down into smaller chunks aimed at specific parts of the object. For example, a separate method for `s.Replicas`, `RolloutStrategy` and so  on.

Reply:
+1 for that suggestion

Reply:
Hm usually we don't break it down that far. Not really against splitting it up, but if was intended as more or less on func per field I think that's a bit much.

@fabriziopandini wdyt?

Reply:
FYI, some of the providers to see to be breaking down the validations to smaller functions. Ref: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/master/api/v1alpha4/azuremachine_validation.go#L31-L59